### PR TITLE
[SPARK-19426][SQL] Custom coalescer for Dataset

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This pr added a new API for `coalesce` in `Dataset`; users can specify the custom coalescer which reduces an input Dataset into fewer partitions. This coalescer implementation is the same with the one in `RDD#coalesce` added in #11865 (SPARK-14042).

This is the rework of #16766.

## How was this patch tested?
Added tests in `DatasetSuite`.

### CODE REVIEW

**Comment by maropu** [comment]
@gatorsmile ok

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala [hunk 1]
```diff
@@ -571,7 +570,8 @@ case class UnionExec(children: Seq[SparkPlan]) extends SparkPlan {
  * current upstream partitions will be executed in parallel (per whatever
  * the current partitioning is).
  */
-case class CoalesceExec(numPartitions: Int, child: SparkPlan) extends UnaryExecNode {
+case class CoalesceExec(numPartitions: Int, child: SparkPlan, coalescer: Option[PartitionCoalescer])
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: Could you add the parm description of `coalescer`? also update function descriptions? Thanks~!

Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: ok!

File: core/src/test/scala/org/apache/spark/rdd/RDDSuite.scala [hunk 2]
```diff
@@ -1185,23 +1194,21 @@ class SizeBasedCoalescer(val maxSize: Int) extends PartitionCoalescer with Seria
...
+          updateGroups()
         }
       } else {
-        if (currentGroup.partitions.size == 0) {
+        if (currentGroup.partitions.isEmpty) {
           addPartition(partition, splitSize)
           index += 1
         } else {
-          updateGroups
+          updateGroups()
```
Comment (gatorsmile) [inline core/src/test/scala/org/apache/spark/rdd/RDDSuite.scala]: All the above changes are not related to this PR, right?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 3]
```diff
@@ -753,6 +753,16 @@ case class Repartition(numPartitions: Int, shuffle: Boolean, child: LogicalPlan)
 }
 
 /**
+ * Returns a new RDD that has at most `numPartitions` partitions. This behavior can be modified by
+ * supplying a `PartitionCoalescer` to control the behavior of the partitioning.
+ */
+case class PartitionCoalesce(numPartitions: Int, coalescer: PartitionCoalescer, child: LogicalPlan)
+  extends UnaryNode {
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Adding new logical nodes also needs the updates in multiple different components. (e.g., Optimizer). 

Is that possible to reuse the existing node `Repartition`?

File: core/src/test/scala/org/apache/spark/rdd/RDDSuite.scala (hunk 2 above)
Comment (maropu) [inline core/src/test/scala/org/apache/spark/rdd/RDDSuite.scala]: Yea, I just left the changes of the original author (probably refactoring stuffs?) ..., better remove this?

Comment (gatorsmile) [inline core/src/test/scala/org/apache/spark/rdd/RDDSuite.scala]: I am fine about this, but it might confuse the others. Maybe just remove them in this PR? You can submit a separate PR later.

Comment (maropu) [inline core/src/test/scala/org/apache/spark/rdd/RDDSuite.scala]: ok, I'll drop these from this pr.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 3 above)
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: yea, I think so. I'll try and plz give me days to do so.

**Comment by maropu** [comment]
@gatorsmile ok, could you check again?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 4]
```diff
@@ -746,8 +746,20 @@ abstract class RepartitionOperation extends UnaryNode {
...
+ * @param coalescer Optional coalescer that an user specifies
  */
-case class Repartition(numPartitions: Int, shuffle: Boolean, child: LogicalPlan)
+case class Repartition(
+    numPartitions: Int,
+    shuffle: Boolean,
+    child: LogicalPlan,
+    coalescer: Option[PartitionCoalescer] = None)
   extends RepartitionOperation {
   require(numPartitions > 0, s"Number of partitions ($numPartitions) must be positive.")
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Add a new require here? 
```
require(!shuffle || coalescer.isEmpty, "Custom coalescer is not allowed for repartition(shuffle=true)")
```

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 5]
```diff
@@ -2662,6 +2662,30 @@ class Dataset[T] private[sql](
...
+   * you like (e.g. one node in the case of numPartitions = 1). To avoid this,
+   * you can call repartition. This will add a shuffle step, but means the
+   * current upstream partitions will be executed in parallel (per whatever
+   * the current partitioning is).
+   *
+   * @group typedrel
+   * @since 2.3.0
+   */
+  def coalesce(numPartitions: Int, userDefinedCoalescer: Option[PartitionCoalescer])
+    : Dataset[T] = withTypedPlan {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]:
```Scala
def coalesce(
    numPartitions: Int,
    userDefinedCoalescer: Option[PartitionCoalescer]): Dataset[T] = withTypedPlan {
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 6]
```diff
@@ -596,14 +596,15 @@ object CollapseProject extends Rule[LogicalPlan] {
...
-    //   otherwise, keep unchanged.
+    // 1) When the top node does not enable the shuffle (i.e., coalesce with no user-specified
+    //   strategy), but the child enables the shuffle. Returns the child node if the last
+    //   numPartitions is bigger; otherwise, keep unchanged.
     // 2) In the other cases, returns the top node with the child's child
-    case r @ Repartition(_, _, child: RepartitionOperation) => (r.shuffle, child.shuffle) match {
-      case (false, true) => if (r.numPartitions >= child.numPartitions) child else r
-      case _ => r.copy(child = child.child)
-    }
+    case r @ Repartition(_, _, child: RepartitionOperation, None) =>
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 4 above)
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: ok

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 6 above)
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: sorry, my bad. Fixed.

**Comment by maropu** [comment]
@gatorsmile ok, fixed.

**Comment by maropu** [comment]
@gatorsmile ping

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 7]
```diff
@@ -596,14 +596,17 @@ object CollapseProject extends Rule[LogicalPlan] {
 object CollapseRepartition extends Rule[LogicalPlan] {
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Please also add new test cases to `CollapseRepartitionSuite` for the changes in this rule.

Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: ok

**Comment by jiangxb1987** [comment]
@maropu any update on this issue?

**Comment by maropu** [comment]
@gatorsmile better to close this pr and the jira?

**Comment by gatorsmile** [comment]
Yeah, maybe close it first. Thanks!

**Comment by maropu** [comment]
ok, thanks!

**Comment by SubhamSinghal** [comment]
I am willing to contribute to the missing pieces in this PR, but need guidance on this.
